dcore/diagnostics.py

```diff
@@ -0,0 +1,12 @@
+from firedrake import assemble, dot, dx
+from math import sqrt
+
+
+class Diagnostics(object):
+    pass
+
+
+class ShallowWaterDiagnostics(Diagnostics):
+
+    def l2(self, f):
+        return sqrt(assemble(dot(f, f)*dx))
```

Comment:
This should either be an `@staticmethod` or an `@classmethod`.  See https://julien.danjou.info/blog/2013/guide-python-static-class-abstract-methods for deciding what to do.